### PR TITLE
21482: Create a Footsteps Script

### PR DESCRIPTION
**Pull request by Triplelexx**
- Added three scripts and test sounds:
    hifi/unpublishedScripts/marketplace/footsteps/footstep.js
    hifi/unpublishedScripts/marketplace/footsteps/footstepCollider.js
    hifi/unpublishedScripts/marketplace/footsteps/createFootstepTestSurface.js

**Test Plan:**
Run footstep.js: When walking on static objects you should hear some default footstep sounds that are selected at random.
The volume of the sounds should vary based on the avatar's size and speed.

Run createFootstepTestSurface: a white box surface should be created in front of the avatar. This box should have a set of sounds defined to sound like a snow covered surface.

**Note:**
Test sounds sourced from freesound.org, courtesy of soundscalpel.com
Edited to be 500% faster
License: Attribution 3.0 Unported (CC BY 3.0)
http://freesound.org/people/soundscalpel.com/sounds/110619/
http://freesound.org/people/soundscalpel.com/sounds/110609/
http://freesound.org/people/soundscalpel.com/sounds/110620/
Added directly into the directory.

### CODE REVIEW

**Comment by hifi-gustavo** [comment]
Mac OS X build is available at: https://deployment.highfidelity.com/jobs/pr-build/label%3Dmacos/2173/HighFidelity-Beta-PR11095-ebf110bdbf2c3311c8419e3bbd15ca120837b3f6.dmg

**Comment by hifi-gustavo** [comment]
Windows build is available at: https://deployment.highfidelity.com/jobs/pr-build/label%3Dwindows/2173/HighFidelity-Beta-PR11095-ebf110bdbf2c3311c8419e3bbd15ca120837b3f6.exe

**Comment by YuppyPlays** [comment]
Testing.

**Comment by YuppyPlays** [comment]
@Triplelexx  
I had this problem with previous PR's regarding finding unpublishedscripts folder for the .js files mentioned, is it possible to get a link to the files if possible to continue.

I can't locate them within the builds installed files.

**Comment by themelissabrown** [comment]
https://raw.githubusercontent.com/Triplelexx/hifi/ebf110bdbf2c3311c8419e3bbd15ca120837b3f6/unpublishedScripts/marketplace/footsteps/createFootstepTestSurface.js

**Comment by YuppyPlays** [comment]
continuing testing, thank you @themelissabrown for the assistance with locating these, "I was lost but now I'm found!" :D 

Will post results shortly. :)

**Comment by YuppyPlays** [comment]
Testing complete:

As per the test plan:
Run footstep.js: When walking on static objects you should hear some default footstep sounds that are selected at random. **Check, works on for example a large created box but does not work on the regular grid floor but I believe thats expected.**
The volume of the sounds should vary based on the avatar's size and speed. **Failed, at first it did keep up with volume and speed but would quickly mess up when I changed avatars size a few times, for example: I was really small but my footsteps were louder than when I previously was of the same size and when going back to very large, it seemed to break the noise and it was very quiet- then back to normal size the speed and volume were off (Slower than the previous sounds speed for each step as well as the volume being lower)**

Run createFootstepTestSurface: a white box surface should be created in front of the avatar. This box should have a set of sounds defined to sound like a snow covered surface. **Failed, unfortunately no sounds were generated when walking on the surface spawned using script**

Just in-case it's useful to this issue, logs attached:
[Logs.zip](https://github.com/highfidelity/hifi/files/1191612/Logs.zip)

**Comment by Triplelexx** [comment]
Sorry, been out today with dead phone.

> Run createFootstepTestSurface: a white box surface should be created in front of the avatar. This box should have a set of sounds defined to sound like a snow covered surface. Failed, unfortunately no sounds were generated when walking on the surface spawned using script

@YuppyPlays 
It seems the folder I stuck everything in is not added to build. The three sound files would need to be in the same folder as the scripts to work.

**Comment by Triplelexx** [comment]
> Check, works on for example a large created box but does not work on the regular grid floor but I believe thats expected.

I've tested the script on the grid floor and it does generate sounds. I had to reset the script after changing domain though. It seems like the issue might be due to changing domain.

**Comment by Triplelexx** [comment]
>  Failed, at first it did keep up with volume and speed but would quickly mess up when I changed avatars size a few times, for example: I was really small but my footsteps were louder than when I previously was of the same size and when going back to very large, it seemed to break the noise and it was very quiet- then back to normal size the speed and volume were off (Slower than the previous sounds speed for each step as well as the volume being lower)

@themelissabrown @misslivirose I am unable to replicate this issue, please confirm whether you'd like me to investigate further.

**Comment by YuppyPlays** [comment]
@Triplelexx 

Repro steps for the issue if required:
 - Opened Sandbox, then interface.
 - Skipped tutorial to get into Sandbox home location (Right click screen whilst in VR, then use controllers to GoTo > Home)
 - Ran Footsteps.js
 - Create > Box, then extend the box to be big enough to walk around (15m x 15m)
 - Walk on the box, hear the sound level at 1:1 scale for avatar.
 - Change size to go much larger (Maximum), walk around, notice the changes
 - Change size again to go smallest, notice the changes.
 - set via "=" key to go normal again, notice changes.
 - Repeat the previous, after a few times it tends to act up with the expected volume/speed.

Hope this helps.

**Comment by Triplelexx** [comment]
@YuppyPlays  Thanks. I'll go through that in a minute.

**Comment by Triplelexx** [comment]
I'm still unable to replicate issue with volume, I've added debugging output e.g "FOOTSTEP: volume: 0.3"
@themelissabrown I could use further instructions on how to proceed.

**Comment by marko8904** [comment]
build this please

**Comment by YuppyPlays** [comment]
@Triplelexx I'm not sure if maybe this would help with the steps, but I forgot to mention the method of which I loaded the .js, I did it via Load URL, not sure if this effects the results in any way (such as it loading data from the internet, possible corruption/loss of data/loading issues?)

**Comment by hifi-gustavo** [comment]
Mac OS X build is available at: https://deployment.highfidelity.com/jobs/pr-build/label%3Dmacos/2193/HighFidelity-Beta-PR11095-022c143bc54350c05ea8faf80644cc739bc3614e.dmg

**Comment by hifi-gustavo** [comment]
Windows build is available at: https://deployment.highfidelity.com/jobs/pr-build/label%3Dwindows/2193/HighFidelity-Beta-PR11095-022c143bc54350c05ea8faf80644cc739bc3614e.exe

**Comment by Triplelexx** [comment]
I'll be pushing another update that should provide for more conclusive testing.

File: unpublishedScripts/marketplace/footsteps/createFootstepTestSurface.js [hunk 1]
```diff
@@ -0,0 +1,57 @@
...
+// Created by Triplelexx on 17/08/01
+// Copyright 2017 High Fidelity, Inc.
+//
+// Distributed under the Apache License, Version 2.0.
+// See the accompanying file LICENSE or http://www.apache.org/licenses/LICENSE-2.0.html
+//
+/* eslint indent: ["error", 4, { "outerIIFEBody": 0 }] */
+
+(function() {
+var snowyFootstepSound1 = Script.resolvePath("snow1.wav");
```
Comment (misslivirose) [inline unpublishedScripts/marketplace/footsteps/createFootstepTestSurface.js]: inner lines within the function should be intended per coding styles

File: unpublishedScripts/marketplace/footsteps/createFootstepTestSurface.js [hunk 2]
```diff
@@ -0,0 +1,57 @@
...
+var snowyFootstepSound2 = Script.resolvePath("snow2.wav");
+var snowyFootstepSound3 = Script.resolvePath("snow3.wav");
+var snowyFootstepSounds = [snowyFootstepSound1, snowyFootstepSound2, snowyFootstepSound3];
+var startPosition = Vec3.sum(Vec3.sum(MyAvatar.position, {
+    x: 0,
+    y: 3 * MyAvatar.scale,
```
Comment (misslivirose) [inline unpublishedScripts/marketplace/footsteps/createFootstepTestSurface.js]: What's the specific reason for 3? Should probably be a named variable

File: unpublishedScripts/marketplace/footsteps/footstep.js [hunk 3]
```diff
@@ -0,0 +1,86 @@
...
+// 
+// Created by Triplelexx on 17/07/27
+// Copyright 2017 High Fidelity, Inc.
+//
+// Distributed under the Apache License, Version 2.0.
+// See the accompanying file LICENSE or http://www.apache.org/licenses/LICENSE-2.0.html
+//
+/* eslint indent: ["error", 4, { "outerIIFEBody": 0 }] */
+
+(function() {
```
Comment (misslivirose) [inline unpublishedScripts/marketplace/footsteps/footstep.js]: indentation within function

File: unpublishedScripts/marketplace/footsteps/createFootstepTestSurface.js (hunk 2 above)
Comment (Triplelexx) [inline unpublishedScripts/marketplace/footsteps/createFootstepTestSurface.js]: I did not intend for the test script to be published. I'm just arbitrarily setting the position of the surface so it doesn't intersect.

**Comment by Triplelexx** [comment]
I will need some more information regarding the entity states to handle edge case issues effectively.
I'll check back in tomorrow.

**Comment by Triplelexx** [comment]
Right, I think I found a solution that works. I think this should pass QA now. I'll do the tidy pass after if so.
The sounds are now hosted. Testing still needs to be done by getting the raw urls from changed files.

**Comment by hifi-gustavo** [comment]
Mac OS X build is available at: https://deployment.highfidelity.com/jobs/pr-build/label%3Dmacos/2259/HighFidelity-Beta-PR11095-e08161dcefa7e4665bc59632afc239ddc7462f36.dmg

**Comment by hifi-gustavo** [comment]
Windows build is available at: https://deployment.highfidelity.com/jobs/pr-build/label%3Dwindows/2259/HighFidelity-Beta-PR11095-e08161dcefa7e4665bc59632afc239ddc7462f36.exe

**Comment by cain-kilgore** [comment]
Testing.

**Comment by cain-kilgore** [comment]
# Test Failed
Test | Result
-------- | --------
Run footstep.js: When walking on static objects you should hear some default footstep sounds that are selected at random. | Fail
The volume of the sounds should vary based on the avatar's size and speed. | Fail
Run createFootstepTestSurface: a white box surface should be created in front of the avatar. This box should have a set of sounds defined to sound like a snow covered surface. | Fail

I was not able to hear any stepping sounds when walking around. Changing my avatar / size / speed did not any difference to this. The white box however was created but did not result in any snow covered surface sounds when walking on it.

Let me know when this needs further testing.

###### Intel i5 7600K @ 4Ghz, 16GB DDR4 2133Mhz Corsair Vengeance, Asus Strix GTX 1050Ti 3GB, HTC Vive

**Comment by Triplelexx** [comment]
Ok, I won't be able to look into anything again for a while. If anyone can provide any further information it would be helpful.

**Comment by themelissabrown** [comment]
We think this is failing because of a bug
https://highfidelity.fogbugz.com/f/cases/edit/1939/Sometimes-collisions-stop-or-never-work-for-an-entity
looking into it more on our side.

**Comment by Triplelexx** [comment]
I also was unable to get collisions to work when seeing the entities to
client only at all.

**Comment by hifi-gustavo** [comment]
Windows build is available at: https://deployment.highfidelity.com/jobs/pr-build/label%3Dwindows/2564/HighFidelity-Beta-PR11095-6605dde0c79226ab6f7065d0d88596e9bdbe176b.exe

**Comment by hifi-gustavo** [comment]
Mac OS X build is available at: https://deployment.highfidelity.com/jobs/pr-build/label%3Dmacos/2564/HighFidelity-Beta-PR11095-6605dde0c79226ab6f7065d0d88596e9bdbe176b.dmg

File: unpublishedScripts/marketplace/footsteps/footstep.js [hunk 4]
```diff
@@ -11,107 +11,106 @@
...
-    var searchSuccess = false;
-    for (var i = 0; i < foundEntities.length; i++){
-        var entityName = Entities.getEntityProperties(foundEntities[i], ["name"]).name;
-        if (entityName == "footstepLGenerator") {
-            searchSuccess = true;
+    function checkFootstepGenerators() {
+        Script.clearInterval(checkGeneratorsInterval);
+        var foundEntities = Entities.findEntities(MyAvatar.position, CHECK_GENERATORS_SEARCH_RADIUS);
+        var searchSuccess = false;
+        for (var i = 0; i < foundEntities.length; i++){
```
Comment (misslivirose) [inline unpublishedScripts/marketplace/footsteps/footstep.js]: space after closing parenthesis and bracket

**Comment by YuppyPlays** [comment]
Testing.

**Comment by YuppyPlays** [comment]
Testing complete:

Test Plan:
Run footstep.js: When walking on static objects you should hear some default footstep sounds that are selected at random. **Pass? - I noticed something I did not before, if the static object is not a certain thickness (I think half a meter it looks like is the minimum?) it causes the sound to not work, however, when above that threshold, it seems to work perfectly. No more issues regarding size changing, though this could have entirely been a one off previously.**
The volume of the sounds should vary based on the avatar's size and speed. **Pass**

Run createFootstepTestSurface: a white box surface should be created in front of the avatar. This box should have a set of sounds defined to sound like a snow covered surface. **Fail?.. - Same as previous in terms of the thickness of the object/cube thats spawned, The white cube with the snow sounds do not seem to work at its spawned in size, Increasing the size via Y scale seems to correct this.**

Please let me know if there is anything more I can do/provide regarding this.

**Comment by Triplelexx** [comment]
@YuppyPlays @cozza13 Regarding the problem described:

>  if the static object is not a certain thickness (I think half a meter it looks like is the minimum?) it causes the sound to not work, however, when above that threshold, it seems to work perfectly. 

The problem does not appear to have anything to do with the thickness of the object.
I tested the script on a surface that was only 0.01m thick and it still works.

It seems to me that there is something happening internally that is stopping collisions from being detected. I did not encounter this issue before, this seems to have become an issue since the start of this job. 
Modifying the scale of objects appears to be resetting collision detection.

I can't see this as having anything to do with my script.
Please advise how to proceed.

**Comment by BillWindwalker** [comment]
,Is Avatar size and position relevant to testing ?

**Comment by Triplelexx** [comment]
I only went so far as to test with an avatar at default size.

**Comment by themelissabrown** [comment]
Trying to figure out a bug on hifi side. Triplelexx will continue work once we fix that.

**Comment by themelissabrown** [comment]
@Triplelexx we're able to repro this bug but only consistently on this script.  Can you review it once more and see if there's anything that jumps out now? If not, we'll need to close the PR.

**Comment by themelissabrown** [comment]
Closing this for now and will re-submit to the hifi-content repo once we figure out why the collision sounds are intermittent.

**Pull request closed without merging**